src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,323 @@
+  std::map<std::string, GridEntry>::iterator it;
+  for (it = grids_.begin(); it != grids_.end(); it++) {
+    std::string typeName;
+    VoxelGridType type = it->second.type;
+    existingGrids.push_back(std::make_pair(it->first, type));
+  }
+  return existingGrids;
+}
+
+std::shared_ptr<Mn::Trade::MeshData> VoxelGrid::getMeshData(
+    const std::string& gridName) {
+  if (meshDataDict_[gridName] == nullptr)
+    generateMesh(gridName);
+  return meshDataDict_[gridName];
+}
+
+Mn::GL::Mesh& VoxelGrid::getMeshGL(const std::string& gridName) {
+  if (meshDataDict_[gridName] == nullptr)
+    generateMesh(gridName);
+  return meshGLDict_[gridName];
+}
+
+Mn::Vector3 VoxelGrid::getGlobalCoords(const Mn::Vector3i& coords) {
+  return Mn::Vector3(coords) * m_voxelSize + m_offset;
+}
+
+void VoxelGrid::fillBoolGridNeighborhood(std::vector<bool>& neighbors,
+                                         const std::string& gridName,
+                                         const Mn::Vector3i& index) {
+  Mn::Vector3i increments[] = {{0, 0, 1},  {1, 0, 0},  {0, 1, 0},
+                               {0, 0, -1}, {0, -1, 0}, {-1, 0, 0}};
+  Cr::Containers::StridedArrayView3D<bool> grid = getGrid<bool>(gridName);
+  for (int i = 0; i < 6; i++) {
+    auto n = index + increments[i];
+    neighbors.push_back(isValidIndex(n) ? grid[n[0]][n[1]][n[2]] : false);
+  }
+}
+
+void VoxelGrid::addVoxelToMeshPrimitives(std::vector<Mn::Vector3>& positions,
+                                         std::vector<Mn::Vector3>& normals,
+                                         std::vector<Mn::Color3>& colors,
+                                         std::vector<Mn::UnsignedInt>& indices,
+                                         const Mn::Vector3i& local_coords,
+                                         const std::vector<bool>& neighbors,
+                                         const Mn::Color3& color) {
+  // Using the data of a cubeSolid to create the voxel cube
+  assert(neighbors.size() >= 6);
+  Mn::Trade::MeshData cubeData = Mn::Primitives::cubeSolid();
+
+  // add cube to mesh
+  // midpoint of a voxel
+  Mn::Vector3 mid = getGlobalCoords(local_coords);
+
+  unsigned int sz = positions.size();
+  auto cubePositions = cubeData.positions3DAsArray();
+  auto cubeNormals = cubeData.normalsAsArray();
+  auto cubeIndices = cubeData.indices();
+
+  // TODO: Only add neccessary faces (based on neighborhood)
+  for (int i = 0; i < 24; i++) {
+    Mn::Vector3 vertOffset = cubePositions[i] * m_voxelSize / 2;
+    positions.push_back(vertOffset + mid);
+    // Set the normals to be weighted such that cubes look slightly curved
+    normals.push_back(cubePositions[i].normalized() * 1 / 4 +
+                      cubeNormals[i].normalized() * 3 / 4);
+    colors.push_back(color);
+  }
+
+  for (int i = 0; i < 6; i++) {
+    if (!neighbors[i]) {
+      for (int j = 0; j < 6; j++) {
+        indices.push_back(sz + cubeIndices[i * 6 + j][0]);
+      }
+    }
+  }
```

Comment:
And this also changed similarly as the Step 4a/4b mentioned before, together with the function signature update.